diff --git a/eagle/ge_data/allocation.py b/eagle/ge_data/allocation.py
index c540794..f2603ee 100644
--- a/eagle/ge_data/allocation.py
+++ b/eagle/ge_data/allocation.py
@@ -13,7 +13,7 @@ e = 68000 - 1
 #e = 68 - 1
 #gpus = [[0],[1],[2],[3],[4],[5],[6],[7]]
 
-gpus=[[0],[1],[2],[3]]
+gpus=[[0, 1],[2, 3]]
 num_p = len(gpus)
 outdir = '{}/sharegpt_{}_{}_mufp16'.format(args.outdir,s,e)
 
@@ -55,7 +55,7 @@ for i in range(num_p):
     gpu_index = gpus[i]
     gpu_index_str = ' '.join(map(str, gpu_index))
     # gpu_index_str='['+gpu_index_str+']'
-    command = "python ge_data_all_vicuna.py --start={} --end={} --index={} --gpu_index {} --outdir {}".format(start, end, index,
+    command = "python ge_data_all_llama2chat.py --start={} --end={} --index={} --gpu_index {} --outdir {}".format(start, end, index,
                                                                                                 gpu_index_str, outdir)
     commands.append(command)
 # run_command(commands[0])
diff --git a/eagle/ge_data/ge_data_all_llama2chat.py b/eagle/ge_data/ge_data_all_llama2chat.py
index eba6161..56b6011 100644
--- a/eagle/ge_data/ge_data_all_llama2chat.py
+++ b/eagle/ge_data/ge_data_all_llama2chat.py
@@ -19,7 +19,7 @@ from datasets import load_dataset
 import json
 from fastchat.model.model_adapter import get_conversation_template
 
-bigname="/home/hongyanz/scratch/weights/llama2chat/13B"
+bigname="meta-llama/Llama-2-7b-chat-hf"
 # bigname = "/home/lyh/weights/hf/llama/7B/"
 # smallname = "/home/lyh/weights/hf/llama/7B/"
 
@@ -43,7 +43,11 @@ def build_dataset_rank(
         tokenizer, split="train",
         select=None,
 ):
-    ds = load_dataset('json', data_files="/home/hongyanz/scratch/data/ShareGPT_V4.3_unfiltered_cleaned_split.json")
+    ds = load_dataset(
+        path="Aeala/ShareGPT_Vicuna_unfiltered",
+        data_files=["ShareGPT_V4.3_unfiltered_cleaned_split.json"],
+        revision='8b0048ad6ae8c22f46a78c15559dec98feef5539'
+    )
     ds = ds['train']
     ds = ds.shuffle(seed=42)
     ds1 = ds.select(range(args.start, args.end))
@@ -142,7 +146,7 @@ def build_dataset_rank(
         batched=True,
         num_proc=num_proc,
         remove_columns=original_columns1,
-        load_from_cache_file=False
+        #load_from_cache_file=False
     )
 
     # ds1 = ds1.filter(lambda x: len(x["input_ids"]) < 1024, batched=False)
